src/data/nat/factorization.lean

```diff
@@ -43,66 +45,122 @@ noncomputable def factorization (n : ℕ) : ℕ →₀ ℕ := (n.factors : multi
-λ a ha b hb h, eq_of_count_factors_eq
-  (zero_lt_iff.mpr ha) (zero_lt_iff.mpr hb) (λ p, by simp [←factorization_eq_count, h])
+λ a ha b hb h, eq_of_count_factors_eq ha hb (λ p, by simp [h])
 
-@[simp] lemma factorization_zero : factorization 0 = 0  :=
+@[simp] lemma factorization_zero : factorization 0 = 0 :=
 by simp [factorization]
 
 @[simp] lemma factorization_one : factorization 1 = 0 :=
 by simp [factorization]
 
 /-- The support of `n.factorization` is exactly `n.factors.to_finset` -/
-@[simp] lemma support_factorization {n : ℕ} :
-  n.factorization.support = n.factors.to_finset :=
+@[simp] lemma support_factorization {n : ℕ} : n.factorization.support = n.factors.to_finset :=
 by simpa [factorization, multiset.to_finsupp_support]
 
 lemma factor_iff_mem_factorization {n p : ℕ} : p ∈ n.factorization.support ↔ p ∈ n.factors :=
 by simp only [support_factorization, list.mem_to_finset]
 
 lemma prime_of_mem_factorization {n p : ℕ} : p ∈ n.factorization.support → p.prime :=
-(@prime_of_mem_factors n p) ∘ (@factor_iff_mem_factorization n p).mp
+prime_of_mem_factors ∘ (@factor_iff_mem_factorization n p).mp
 
 lemma pos_of_mem_factorization {n p : ℕ} : p ∈ n.factorization.support → 0 < p :=
-(@prime.pos p) ∘ (@prime_of_mem_factorization n p)
+prime.pos ∘ (@prime_of_mem_factorization n p)
 
 /-- The only numbers with empty prime factorization are `0` and `1` -/
 lemma factorization_eq_zero_iff (n : ℕ) : n.factorization = 0 ↔ n = 0 ∨ n = 1 :=
 by simp [factorization, add_equiv.map_eq_zero_iff, multiset.coe_eq_zero]
 
+/-- For positive `a` and `b`, the power of `p` in `a * b` is the sum of the powers in `a` and `b` -/
+lemma count_factors_mul {p a b : ℕ} (ha : a ≠ 0) (hb : b ≠ 0) :
+  list.count p (a * b).factors = list.count p a.factors + list.count p b.factors :=
+by rw [perm_iff_count.mp (perm_factors_mul ha hb) p, count_append]
+
 /-- For nonzero `a` and `b`, the power of `p` in `a * b` is the sum of the powers in `a` and `b` -/
 @[simp] lemma factorization_mul {a b : ℕ} (ha : a ≠ 0) (hb : b ≠ 0) :
   (a * b).factorization = a.factorization + b.factorization :=
-by { ext p, simp only [add_apply, factorization_eq_count,
-  count_factors_mul_of_pos (zero_lt_iff.mpr ha) (zero_lt_iff.mpr hb)] }
+by { ext p, simp only [add_apply, ←factors_count_eq, count_factors_mul ha hb] }
 
 /-- For any `p`, the power of `p` in `n^k` is `k` times the power in `n` -/
-lemma factorization_pow {n k : ℕ} :
+@[simp] lemma factorization_pow (n k : ℕ) :
   factorization (n^k) = k • n.factorization :=
-by { ext p, simp [factorization_eq_count, factors_count_pow] }
+begin
+  induction k with k ih,
+  { simp },
+  rcases n.eq_zero_or_pos with rfl | hn,
+  { simp },
+  rw [pow_succ, factorization_mul hn.ne' (pow_ne_zero _ hn.ne'), ih, succ_eq_one_add, add_smul,
+    one_smul],
+end
+
+/-- For coprime `a` and `b`, the power of `p` in `a * b` is the sum of the powers in `a` and `b` -/
+lemma count_factors_mul_of_coprime {p a b : ℕ} (hab : coprime a b)  :
```

Comment:
by the way, should we keep lemmas like this around or should they go the way of the dinosaur?

Reply:
Do you mean ones which are about `coprime` rather than `pos`, or ones about `factors.count`?

Reply:
`factors.count`

Reply:
Yeah I think they should all die, and we should have just things about `factorization` - I killed some of them already but I think I just moved this one for now with the view of replacing it later? Not sure though, feel free to change! In my mind the two important factors are that: no lemma mentions `factors.count` except for the one saying it's the same as `factorization`; and we haven't lost any functionality in the overall API.